File: module-infra/src/main/java/maple/expectation/infrastructure/resilience/CompensationLogService.java
package maple.expectation.infrastructure.resilience;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.error.exception.MapleDataProcessingException;
import maple.expectation.infrastructure.executor.CheckedLogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import maple.expectation.infrastructure.executor.function.CheckedRunnable;
import maple.expectation.infrastructure.executor.function.CheckedSupplier;
import org.redisson.api.RStream;
import org.redisson.api.RedissonClient;
import org.redisson.api.StreamMessageId;
import org.redisson.api.stream.StreamAddArgs;
import org.redisson.api.stream.StreamCreateGroupArgs;
import org.redisson.api.stream.StreamReadGroupArgs;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;

/**
 * Compensation Log Service (Issue #218)
 *
 * <p>MySQL 장애 중 발생한 데이터 변경을 Redis Stream에 기록합니다.
 *
 * <h4>Redis Stream 구조 (P1-2 Hash Tag)</h4>
 *
 * <ul>
 *   <li>Stream: {mysql}:compensation:stream
 *   <li>DLQ: {mysql}:compensation:dlq
 *   <li>Consumer Group: compensation-sync
 * </ul>
 *
 * <h4>엔트리 필드</h4>
 *
 * <ul>
 *   <li>type: 데이터 타입 (equipment, ocid)
 *   <li>key: 캐시 키
 *   <li>data: JSON 직렬화된 데이터
 *   <li>timestamp: 기록 시각
 *   <li>retryCount: 재시도 횟수
 * </ul>
 */
@Slf4j
@Service
@RequiredArgsConstructor
@ConditionalOnProperty(
    prefix = "resilience.mysql-fallback",
    name = "enabled",
    havingValue = "true",
    matchIfMissing = false)
public class CompensationLogService {

  private static final String FIELD_TYPE = "type";
  private static final String FIELD_KEY = "key";
  private static final String FIELD_DATA = "data";
  private static final String FIELD_TIMESTAMP = "timestamp";
  private static final String FIELD_RETRY_COUNT = "retryCount";

  private final RedissonClient redissonClient;
  private final MySQLFallbackProperties properties;
  private final ObjectMapper objectMapper;
  private final CheckedLogicExecutor checkedExecutor;

  private final String instanceId = generateInstanceId();

  private static String generateInstanceId() {
    return Optional.ofNullable(System.getenv("HOSTNAME"))
            .or(() -> Optional.ofNullable(System.getenv("COMPUTERNAME")))
            .orElse("unknown")
        + "-"
        + UUID.randomUUID().toString().substring(0, 8);
  }

  /** Consumer Group 초기화 (P1-N1) */
  @PostConstruct
  public void initConsumerGroup() {
    log.info("[CompensationLog] Instance consumer ID: {}", instanceId);

    checkedExecutor.executeUncheckedVoid(
        new CheckedRunnable() {
          @Override
          public void run() throws Exception {
            RStream<String, String> stream =
                redissonClient.getStream(properties.getCompensationStream());

            // 스트림이 없으면 생성 및 그룹 생성
            if (!stream.isExists()) {
              stream.createGroup(
                  StreamCreateGroupArgs.name(properties.getSyncConsumerGroup()).makeStream());
              log.info(
                  "[CompensationLog] Stream 및 Consumer Group 생성: {}",
                  properties.getSyncConsumerGroup());
              return;
            }

            // 스트림이 존재하면 그룹 존재 여부 확인 후 생성 (Redis snapshot restore 대응)
            try {
              stream.readGroup(
                  properties.getSyncConsumerGroup(),
                  instanceId,
                  StreamReadGroupArgs.neverDelivered().count(1));
            } catch (Exception e) {
              // NOGROUP 에러인 경우 그룹 생성
              if (e.getMessage() != null && e.getMessage().contains("NOGROUP")) {
                stream.createGroup(StreamCreateGroupArgs.name(properties.getSyncConsumerGroup()));
                log.info(
                    "[CompensationLog] Consumer Group 재생성 (NOGROUP 복구): {}",
                    properties.getSyncConsumerGroup());
              } else {
                throw e;
              }
            }
          }
        },
        TaskContext.of("Compensation", "InitConsumerGroup", properties.getSyncConsumerGroup()),
        e -> new MapleDataProcessingException("Consumer Group 초기화 실패", e));
  }

  /**
   * Compensation Log 기록 (P1-N3: MAXLEN 적용)
   *
   * @param type 데이터 타입 (equipment, ocid)
   * @param key 캐시 키
   * @param data 저장할 데이터 객체
   * @return Stream 메시지 ID
   */
  public StreamMessageId writeLog(String type, String key, Object data) {
    return checkedExecutor.executeUnchecked(
        new CheckedSupplier<StreamMessageId>() {
          @Override
          public StreamMessageId get() throws Exception {
            String jsonData = serializeDataSafely(data);

            RStream<String, String> stream =
                redissonClient.getStream(properties.getCompensationStream());

            // XADD with MAXLEN ~ (P1-N3: trimNonStrict = approximate ~)
            StreamMessageId messageId =
                stream.add(
                    StreamAddArgs.entries(
                            Map.of(
                                FIELD_TYPE, type,
                                FIELD_KEY, key,
                                FIELD_DATA, jsonData,
                                FIELD_TIMESTAMP, Instant.now().toString(),
                                FIELD_RETRY_COUNT, "0"))
                        .trimNonStrict()
                        .maxLen(properties.getStreamMaxLen())
                        .noLimit());

            log.info(
                "[CompensationLog] 로그 기록 완료: type={}, key={}, messageId={}", type, key, messageId);
            return messageId;
          }
        },
        TaskContext.of("Compensation", "WriteLog", key),
        e -> new MapleDataProcessingException("Compensation Log 기록 실패: " + key, e));
  }

  /**
   * Compensation Log 읽기 (Consumer Group)
   *
   * <p>P1 Fix: Pending 메시지를 먼저 확인하고 재시도 후 새 메시지를 읽습니다.
   *
   * <p>1. XAUTOCLAIM로 10분 이상 대기 중인 메시지 확인 및 자동 CLAIM (crashed consumer 감지)
   *
   * <p>2. Claim된 메시지가 있으면 반환
   *
   * <p>3. Claim된 메시지가 없으면 새 메시지 읽기
   *
   * @param consumerId Consumer 식별자 (deprecated, instanceId 사용됨)
   * @param count 읽을 최대 메시지 수
   * @return 읽은 메시지 목록
   */
  public Map<StreamMessageId, Map<String, String>> readLogs(String consumerId, int count) {
    return checkedExecutor.executeUnchecked(
        new CheckedSupplier<Map<StreamMessageId, Map<String, String>>>() {
          @Override
          public Map<StreamMessageId, Map<String, String>> get() throws Exception {
            RStream<String, String> stream =
                redissonClient.getStream(properties.getCompensationStream());

            // P1 Fix: Pending 메시지 확인 및 자동 CLAIM (crashed consumer 재시도)
            // 10분(600,000ms) 이상 delivery된 메시지만 CLAIM (타 consumer가 처리 중인 메시지 방지)
            var autoClaimResult =
                stream.autoClaim(
                    properties.getSyncConsumerGroup(),
                    instanceId,
                    600_000, // min idle time: 10 minutes
                    TimeUnit.MILLISECONDS,
                    StreamMessageId.MIN, // start from beginning
                    count);

            Map<StreamMessageId, Map<String, String>> claimedMessages =
                autoClaimResult.getMessages();
            if (!claimedMessages.isEmpty()) {
              log.info("[CompensationLog] Pending 메시지 재처리: {} 건", claimedMessages.size());
              return claimedMessages;
            }

            // Pending 메시지가 없으면 새 메시지 읽기
            return stream.readGroup(
                properties.getSyncConsumerGroup(),
                instanceId,
                StreamReadGroupArgs.neverDelivered().count(count));
          }
        },
        TaskContext.of("Compensation", "ReadLogs", instanceId),
        e -> new MapleDataProcessingException("Compensation Log 읽기 실패", e));
  }

  /**
   * Compensation Log ACK (처리 완료)
   *
   * @param messageIds 완료된 메시지 ID 목록
   */
  public void ackLogs(List<StreamMessageId> messageIds) {
    if (messageIds.isEmpty()) {
      return;
    }

    checkedExecutor.executeUncheckedVoid(
        new CheckedRunnable() {
          @Override
          public void run() throws Exception {
            RStream<String, String> stream =
                redissonClient.getStream(properties.getCompensationStream());
            stream.ack(
                properties.getSyncConsumerGroup(), messageIds.toArray(new StreamMessageId[0]));
            log.debug("[CompensationLog] ACK 완료: {} 건", messageIds.size());
          }
        },
        TaskContext.of("Compensation", "AckLogs", String.valueOf(messageIds.size())),
        e -> new MapleDataProcessingException("Compensation Log ACK 실패", e));
  }

  /**
   * DLQ로 이동 (P1-4: 3회 실패 후)
   *
   * @param originalMessageId 원본 메시지 ID
   * @param entry 원본 메시지 데이터
   * @param errorMessage 실패 원인
   */
  public void moveToDlq(
      StreamMessageId originalMessageId, Map<String, String> entry, String errorMessage) {
    checkedExecutor.executeUncheckedVoid(
        new CheckedRunnable() {
          @Override
          public void run() throws Exception {
            RStream<String, String> dlqStream =
                redissonClient.getStream(properties.getCompensationDlq());

            // DLQ에 에러 정보 추가하여 기록 (P1-N3: trimNonStrict = approximate ~)
            dlqStream.add(
                StreamAddArgs.entries(
                        Map.of(
                            FIELD_TYPE,
                            entry.getOrDefault(FIELD_TYPE, "unknown"),
                            FIELD_KEY,
                            entry.getOrDefault(FIELD_KEY, "unknown"),
                            FIELD_DATA,
                            entry.getOrDefault(FIELD_DATA, "{}"),
                            FIELD_TIMESTAMP,
                            entry.getOrDefault(FIELD_TIMESTAMP, Instant.now().toString()),
                            FIELD_RETRY_COUNT,
                            entry.getOrDefault(FIELD_RETRY_COUNT, "0"),
                            "originalMessageId",
                            originalMessageId.toString(),
                            "errorMessage",
                            errorMessage,
                            "dlqTimestamp",
                            Instant.now().toString()))
                    .trimNonStrict()
                    .maxLen(properties.getStreamMaxLen())
                    .noLimit());

            // 원본 스트림에서 ACK (재처리 방지)
            ackLogs(List.of(originalMessageId));

            log.warn(
                "[CompensationLog] DLQ 이동: messageId={}, error={}",
                originalMessageId,
                errorMessage);
          }
        },
        TaskContext.of("Compensation", "MoveToDlq", originalMessageId.toString()),
        e -> new MapleDataProcessingException("DLQ 이동 실패: " + originalMessageId, e));
  }

  /** Pending 메시지 수 조회 (메트릭용) */
  public long getPendingCount() {
    return checkedExecutor.executeUnchecked(
        new CheckedSupplier<Long>() {
          @Override
          public Long get() throws Exception {
            RStream<String, String> stream =
                redissonClient.getStream(properties.getCompensationStream());
            return stream.getPendingInfo(properties.getSyncConsumerGroup()).getTotal();
          }
        },
        TaskContext.of("Compensation", "GetPendingCount", properties.getCompensationStream()),
        e -> new MapleDataProcessingException("Pending 메시지 수 조회 실패", e));
  }

  /** DLQ 메시지 수 조회 (메트릭용) */
  public long getDlqCount() {
    return checkedExecutor.executeUnchecked(
        new CheckedSupplier<Long>() {
          @Override
          public Long get() throws Exception {
            RStream<String, String> dlqStream =
                redissonClient.getStream(properties.getCompensationDlq());
            return dlqStream.isExists() ? dlqStream.size() : 0;
          }
        },
        TaskContext.of("Compensation", "GetDlqCount", properties.getCompensationDlq()),
        e -> new MapleDataProcessingException("DLQ 메시지 수 조회 실패", e));
  }

  /** JSON 직렬화 */
  private String serializeDataSafely(Object data) {
    try {
      return objectMapper.writeValueAsString(data);
    } catch (JsonProcessingException e) {
      throw new MapleDataProcessingException("JSON 직렬화 실패: " + data.getClass().getSimpleName(), e);
    }
  }

  /** JSON 역직렬화 */
  public <T> T deserializeData(String json, Class<T> type) {
    return checkedExecutor.executeUnchecked(
        new CheckedSupplier<T>() {
          @Override
          public T get() throws Exception {
            try {
              return objectMapper.readValue(json, type);
            } catch (JsonProcessingException e) {
              throw new MapleDataProcessingException("JSON 역직렬화 실패: " + type.getSimpleName(), e);
            }
          }
        },
        TaskContext.of("Compensation", "DeserializeData", type.getSimpleName()),
        e -> new MapleDataProcessingException("JSON 역직렬화 실패: " + type.getSimpleName(), e));
  }

  /** JSON 역직렬화 안전 처리 */
  private <T> T deserializeDataSafely(String json, Class<T> type) throws JsonProcessingException {
    return objectMapper.readValue(json, type);
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/resilience/CompensationLogService.kt
package maple.expectation.infrastructure.resilience

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.annotation.PostConstruct
import java.time.Instant
import java.util.Optional
import java.util.UUID
import java.util.concurrent.TimeUnit
import maple.expectation.error.exception.MapleDataProcessingException
import maple.expectation.infrastructure.executor.CheckedLogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import org.redisson.api.RStream
import org.redisson.api.RedissonClient
import org.redisson.api.StreamMessageId
import org.redisson.api.stream.StreamAddArgs
import org.redisson.api.stream.StreamCreateGroupArgs
import org.redisson.api.stream.StreamReadGroupArgs
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service

@Service
@ConditionalOnProperty(
    prefix = "resilience.mysql-fallback",
    name = ["enabled"],
    havingValue = "true",
    matchIfMissing = false,
)
class CompensationLogService(
    private val redissonClient: RedissonClient,
    private val properties: MySQLFallbackProperties,
    private val objectMapper: ObjectMapper,
    private val checkedExecutor: CheckedLogicExecutor,
) {
    private val logger = LoggerFactory.getLogger(CompensationLogService::class.java)

    private val instanceId: String = generateInstanceId()

    companion object {
        private const val FIELD_TYPE = "type"
        private const val FIELD_KEY = "key"
        private const val FIELD_DATA = "data"
        private const val FIELD_TIMESTAMP = "timestamp"
        private const val FIELD_RETRY_COUNT = "retryCount"

        private fun generateInstanceId(): String = Optional.ofNullable(System.getenv("HOSTNAME"))
            .or { Optional.ofNullable(System.getenv("COMPUTERNAME")) }
            .orElse("unknown") + "-" + UUID.randomUUID().toString().substring(0, 8)
    }

    @PostConstruct
    fun initConsumerGroup() {
        logger.info("[CompensationLog] Instance consumer ID: {}", instanceId)

        checkedExecutor.executeUncheckedVoid(
            {
                val stream = redissonClient.getStream<Any, Any>(properties.compensationStream)

                if (!stream.isExists) {
                    stream.createGroup(StreamCreateGroupArgs.name(properties.syncConsumerGroup).makeStream())
                    logger.info(
                        "[CompensationLog] Stream 및 Consumer Group 생성: {}",
                        properties.syncConsumerGroup,
                    )
                    return@executeUncheckedVoid
                }

                try {
                    stream.readGroup(
                        properties.syncConsumerGroup,
                        instanceId,
                        StreamReadGroupArgs.neverDelivered().count(1),
                    )
                } catch (e: Exception) {
                    if (e.message?.contains("NOGROUP") == true) {
                        stream.createGroup(StreamCreateGroupArgs.name(properties.syncConsumerGroup))
                        logger.info(
                            "[CompensationLog] Consumer Group 재생성 (NOGROUP 복구): {}",
                            properties.syncConsumerGroup,
                        )
                    } else {
                        throw e
                    }
                }
            },
            TaskContext.of("Compensation", "InitConsumerGroup", properties.syncConsumerGroup),
        ) { e: Throwable -> MapleDataProcessingException("Consumer Group 초기화 실패", e) }
    }

    fun writeLog(type: String, key: String, data: Any): StreamMessageId = checkedExecutor.executeUnchecked(
        {
            val jsonData = serializeDataSafely(data)

            val stream: RStream<Any, Any> = redissonClient.getStream(properties.compensationStream)

            val messageId = stream.add(
                StreamAddArgs.entries<Any, Any>(
                    mapOf(
                        FIELD_TYPE to type,
                        FIELD_KEY to key,
                        FIELD_DATA to jsonData,
                        FIELD_TIMESTAMP to Instant.now().toString(),
                        FIELD_RETRY_COUNT to "0",
                    ),
                )
                    .trimNonStrict()
                    .maxLen(properties.streamMaxLen)
                    .noLimit(),
            )

            logger.info(
                "[CompensationLog] 로그 기록 완료: type={}, key={}, messageId={}",
                type,
                key,
                messageId,
            )
            messageId
        },
        TaskContext.of("Compensation", "WriteLog", key),
    ) { e: Throwable -> MapleDataProcessingException("Compensation Log 기록 실패: $key", e) }

    fun readLogs(consumerId: String, count: Int): Map<StreamMessageId, Map<String, String>> {
        return checkedExecutor.executeUnchecked(
            {
                val stream: RStream<Any, Any> = redissonClient.getStream(properties.compensationStream)

                val autoClaimResult = stream.autoClaim(
                    properties.syncConsumerGroup,
                    instanceId,
                    600_000L,
                    TimeUnit.MILLISECONDS,
                    StreamMessageId.MIN,
                    count,
                )

                val claimedMessages = autoClaimResult.messages
                if (claimedMessages.isNotEmpty()) {
                    logger.info("[CompensationLog] Pending 메시지 재처리: {} 건", claimedMessages.size)
                    @Suppress("UNCHECKED_CAST")
                    return@executeUnchecked convertToStringMap(claimedMessages)
                }

                @Suppress("UNCHECKED_CAST")
                val rawMessages = stream.readGroup(
                    properties.syncConsumerGroup,
                    instanceId,
                    StreamReadGroupArgs.neverDelivered().count(count),
                ) as Map<StreamMessageId, Map<Any, Any>>
                convertToStringMap(rawMessages)
            },
            TaskContext.of("Compensation", "ReadLogs", instanceId),
        ) { e: Throwable -> MapleDataProcessingException("Compensation Log 읽기 실패", e) }
    }

    @Suppress("UNCHECKED_CAST")
    private fun convertToStringMap(messages: Map<StreamMessageId, Map<Any, Any>>): Map<StreamMessageId, Map<String, String>> = messages.mapValues { entry ->
        entry.value.mapKeys { it.key as String }
            .mapValues { it.value as String }
    }

    fun ackLogs(messageIds: List<StreamMessageId>) {
        if (messageIds.isEmpty()) {
            return
        }

        checkedExecutor.executeUncheckedVoid(
            {
                val stream = redissonClient.getStream<Any, Any>(properties.compensationStream)
                stream.ack(properties.syncConsumerGroup, *messageIds.toTypedArray())
                logger.debug("[CompensationLog] ACK 완료: {} 건", messageIds.size)
            },
            TaskContext.of("Compensation", "AckLogs", messageIds.size.toString()),
        ) { e: Throwable -> MapleDataProcessingException("Compensation Log ACK 실패", e) }
    }

    fun moveToDlq(originalMessageId: StreamMessageId, entry: Map<String, String>, errorMessage: String) {
        checkedExecutor.executeUncheckedVoid(
            {
                val dlqStream: RStream<Any, Any> = redissonClient.getStream(properties.compensationDlq)

                dlqStream.add(
                    StreamAddArgs.entries<Any, Any>(
                        mapOf(
                            FIELD_TYPE to entry.getOrDefault(FIELD_TYPE, "unknown"),
                            FIELD_KEY to entry.getOrDefault(FIELD_KEY, "unknown"),
                            FIELD_DATA to entry.getOrDefault(FIELD_DATA, "{}"),
                            FIELD_TIMESTAMP to entry.getOrDefault(FIELD_TIMESTAMP, Instant.now().toString()),
                            FIELD_RETRY_COUNT to entry.getOrDefault(FIELD_RETRY_COUNT, "0"),
                            "originalMessageId" to originalMessageId.toString(),
                            "errorMessage" to errorMessage,
                            "dlqTimestamp" to Instant.now().toString(),
                        ),
                    )
                        .trimNonStrict()
                        .maxLen(properties.streamMaxLen)
                        .noLimit(),
                )

                ackLogs(listOf(originalMessageId))

                logger.warn(
                    "[CompensationLog] DLQ 이동: messageId={}, error={}",
                    originalMessageId,
                    errorMessage,
                )
            },
            TaskContext.of("Compensation", "MoveToDlq", originalMessageId.toString()),
        ) { e: Throwable -> MapleDataProcessingException("DLQ 이동 실패: $originalMessageId", e) }
    }

    fun getPendingCount(): Long = checkedExecutor.executeUnchecked(
        {
            val stream = redissonClient.getStream<Any, Any>(properties.compensationStream)
            stream.getPendingInfo(properties.syncConsumerGroup).total
        },
        TaskContext.of("Compensation", "GetPendingCount", properties.compensationStream),
    ) { e: Throwable -> MapleDataProcessingException("Pending 메시지 수 조회 실패", e) }

    fun getDlqCount(): Long = checkedExecutor.executeUnchecked(
        {
            val dlqStream = redissonClient.getStream<Any, Any>(properties.compensationDlq)
            if (dlqStream.isExists) dlqStream.size() else 0L
        },
        TaskContext.of("Compensation", "GetDlqCount", properties.compensationDlq),
    ) { e: Throwable -> MapleDataProcessingException("DLQ 메시지 수 조회 실패", e) }

    private fun serializeDataSafely(data: Any): String = try {
        objectMapper.writeValueAsString(data)
    } catch (e: JsonProcessingException) {
        throw MapleDataProcessingException("JSON 직렬화 실패: ${data.javaClass.simpleName}", e)
    }

    fun <T> deserializeData(json: String, type: Class<T>): T = checkedExecutor.executeUnchecked(
        {
            try {
                objectMapper.readValue(json, type)
            } catch (e: JsonProcessingException) {
                throw MapleDataProcessingException("JSON 역직렬화 실패: ${type.simpleName}", e)
            }
        },
        TaskContext.of("Compensation", "DeserializeData", type.simpleName),
    ) { e: Throwable -> MapleDataProcessingException("JSON 역직렬화 실패: ${type.simpleName}", e) }
}
